First implementation of symfony fixture loader + group support #461

arendjantetteroo opened the pull request:
A first implementation, basically copy/paste from the fixturesbundle. 

Couple questions:
- Do we want to keep depending on the fixturesBundle or should we copy any files over so we can avoid the fixturesBundle which also installs the orm stuff which if you only use odm you don't need? (for example the FixtureGroupInterface)
- i based it on 3.5.3 as that is what i have locally, it would be nice if this could be included in 3.6, but not sure we can keep it backward compatible? Maybe if we keep fixtures/bundle options and keep their old implementation and if you provide an --autowire option it uses the new one? 

Thoughts?

Fixes #461

alcaeus commented on DependencyInjection/Compiler/FixturesCompilerPass.php:
```diff
@@ -0,0 +1,37 @@
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
```
Useless docblock

alcaeus commented on DependencyInjection/Compiler/FixturesCompilerPass.php:
```diff
@@ -0,0 +1,37 @@
+            $definition->addMethodCall('addFixture', [new Reference($serviceId)]);
+        }
+    }
+}
```
There's a missing newline here.

alcaeus commented:
Sorry for the close/reopen spam - travis-ci needed another nudge to check this PR.

> Do we want to keep depending on the fixturesBundle or should we copy any files over so we can avoid the fixturesBundle which also installs the orm stuff which if you only use odm you don't need? (for example the FixtureGroupInterface)

We don't depend on the fixtures bundle, only on the data-fixtures library, which doesn't have a hard requirement on ORM. Since we don't want to introduce one, copying the necessary files from the fixtures bundle is the way to go for now. I'm afraid I won't be able to implement #509 since it most likely requires changes to the fixtures bundle as well.

> i based it on 3.5.3 as that is what i have locally, it would be nice if this could be included in 3.6, but not sure we can keep it backward compatible? Maybe if we keep fixtures/bundle options and keep their old implementation and if you provide an --autowire option it uses the new one?

If there's going to be a 3.6, it will be in there. 3.6 will happen if I find things dropped in 4.0 that I haven't deprecated in 3.5 yet (I remember there were some). I'll give you an update once I've checked that.

I'll do a comprehensive code review shortly, just wanted to give you a quick update since you opened a pull request so quickly! 🎉

alcaeus commented:
And a quick update right behind it: there will definitely be a 3.6 release which we'll get going shortly. Once that's done you can change the base on the PR to 3.6 and it will be there 👍

alcaeus commented on Command/LoadDataFixturesDoctrineODMCommand.php:
```diff
@@ -82,37 +84,18 @@ protected function execute(InputInterface $input, OutputInterface $output)
+        if (! $fixtures) {
+            $message = 'Could not find any fixture services to load';
 
-        $loaderClass = $this->container->getParameter('doctrine_mongodb.odm.fixture_loader');
```
Dropping this is a BC break: we can change the default value in 3.x but not drop it entirely.

arendjantetteroo replied:
@alcaeus 
Should i add an --autowire option to the command, which if provided it uses the new symfonyLoader?

And if not provided, it uses the old way with the --bundles or --fixtures option? 
And a check that if you pass --autowire, you can't pass --bundles or --fxitures?
And add an @deprecated if you use the old 2? 

Then in 4.0 you can remove the --bundles,--fixtures and old loader?

alcaeus replied:
The more I think about it, the more I think we need to move this into 4.0. Then you can break BC as much as you want...

arendjantetteroo replied:
The problem is that i can't easily move to 4.0 as lots of code breaks due to the odm2.0 chnges. So if there is a way to get this onto 3.x i can already clean up my fixtures. 
I'll see if i can find an easy way to have both approaches working

arendjantetteroo replied:
Or maybe add a new fixture loader command and deprecate the old one, where in 4.0 you can only use the new one?

arendjantetteroo commented:
> > Do we want to keep depending on the fixturesBundle or should we copy any files over so we can avoid the fixturesBundle which also installs the orm stuff which if you only use odm you don't need? (for example the FixtureGroupInterface)
> 
> We don't depend on the fixtures bundle, only on the data-fixtures library, which doesn't have a hard requirement on ORM. Since we don't want to introduce one, copying the necessary files from the fixtures bundle is the way to go for now. I'm afraid I won't be able to implement #509 since it most likely requires changes to the fixtures bundle as well.

I copied the Fixture class + the fixtureGroupinterface so it works without the FixturesBundle now. 
So it only depends on the doctrine/data-fixtures library.

arendjantetteroo commented:
I fixed the issue with the tests by adding a mocked loader, the functionality of the loader is not tested. 
I saw that the FixturesBundle has an integration test for it, should we copy that as well?

arendjantetteroo commented:
While fixing the command tests i figured out the current loader code is php 7.1 and up with type checking. 

2 options: 
1. Remove all type stuff so 5.6 is still valid (assuming 3.6 should still be 5.x compatible?)
2. Postpone to 4.0

alcaeus commented:
I'm very inclined to postpone to 4.0 - the trouble doesn't seem to be worth it.

arendjantetteroo commented:
@alcaeus 
After a few more changes, i've reverted the old way in the command with --bundles and --fixtures. 

If a user wants to use the autowired way, he can use --services. 
The old and new options are incompatible, so if both --bundles/--fixtures and --services are used, an error is thrown indicating it. 
if --services is provided, we assume the user knows that he needs to use the fixtures as services. 
I guess this should be added to the 3.6 upgrade file?
In 4.0 you could then remove the --bundles and --fixtures options if you want to and make this only based on services and the --group feature to define which to load. 

It's working on both my old application way of doing fixtures and working with the new autowired fixtures in a new branch. 

Let me know what else is needed?

arendjantetteroo commented on Command/LoadDataFixturesDoctrineODMCommand.php:
```diff
@@ -45,32 +51,41 @@ protected function configure()
-        if ($bundles && $dirOrFile) {
-            throw new \InvalidArgumentException('Use only one option: --bundles or --fixtures.');
+        $services = $input->getOption('services');
+        if ($bundles && $dirOrFile && $services) {
```
This check should be changed so it checks if more than 1 of the three is used.

alcaeus commented:
@arendjantetteroo I'll take a deeper look at the PR tonight. If we can keep this entirely BC I'm all for it.

alcaeus commented on Command/LoadDataFixturesDoctrineODMCommand.php:
```diff
@@ -45,32 +56,50 @@ protected function configure()
+  <info>php %command.full_name%</info>
 
 You can also optionally specify the path to fixtures with the <info>--fixtures</info> option:
+  <info>php %command.full_name%</info> --fixtures=/path/to/fixtures1 --fixtures=/path/to/fixtures2</info>
```
I would keep original documentation intact except for adding a deprecation notice and add a paragraph below to explain new features. Suggested wording:

> Alternatively, you can also load fixture services instead of files. Fixture services are tagged with `<comment>doctrine.fixture.odm</comment>`.
When using `<comment>--services</comment>`, both `<comment>--fixtures</comment>` and `<comment>--bundles</comment>` will no longer work. Using `<comment>--services</comment>` will be the default behaviour in 4.0.
When loading fixture services, you can also choose to load only fixtures that live in a certain group:
`<info>php %command.full_name%</info> <comment>--group=group1</comment> <comment>--services</comment>`

alcaeus commented on Command/LoadDataFixturesDoctrineODMCommand.php:
```diff
@@ -45,32 +56,50 @@ protected function configure()
-        if ($bundles && $dirOrFile) {
-            throw new \InvalidArgumentException('Use only one option: --bundles or --fixtures.');
+        $services = $input->getOption('services');
+        if ($services === false) {
```
I'd simplify the entire block:
```
if (($services && $bundles) || ($services && $dirOrFile) || ($bundles && $dirOrFile)) {
    throw new \InvalidArgumentException('Can only use one of "--bundles", "--fixtures", "--services".');
}
```

I don't think treating `--services` in combination with `--bundles` or `--fixtures` warrants a warning, I'd just throw a straight up error messages: chances are that people who decide to use `--services` already know what it does.

alcaeus commented on Command/LoadDataFixturesDoctrineODMCommand.php:
```diff
@@ -82,37 +111,51 @@ protected function execute(InputInterface $input, OutputInterface $output)
-            $paths = [$this->getKernel()->getRootDir().'/DataFixtures/MongoDB'];
-            foreach ($bundles as $bundle) {
-                $paths[] = $this->getKernel()->getBundle($bundle)->getPath();
+        if ($services === false) {
```
```suggestion
        if (!$services) {
```

alcaeus replied:
This code path also needs to trigger a deprecation notice explaining that using `--bundles` and `--fixtures` is deprecated and will be removed in 4.0.

alcaeus commented on Command/LoadDataFixturesDoctrineODMCommand.php:
```diff
@@ -45,32 +56,50 @@ protected function configure()
         $bundles = $input->getOption('bundles');
-        if ($bundles && $dirOrFile) {
-            throw new \InvalidArgumentException('Use only one option: --bundles or --fixtures.');
+        $services = $input->getOption('services');
```
```suggestion
        $services = (bool) $input->getOption('services');
```

alcaeus commented on Fixture/FixtureGroupInterface.php:
```diff
@@ -0,0 +1,17 @@
+/**
+ * FixtureGroupInterface can be implemented by fixtures that belong in groups
+ */
+interface FixtureGroupInterface
```
I'm not excited to introduce a separate interface for this here, as it means that people that want to use their fixtures for ORM and ODM will have to use two different marker interfaces that provide the same functionality. However, since the corresponding ORM interface resides in the bundle and the library itself doesn't know anything about groups, there might not be a clean way around this 🤔 

I'll have to give this some more thought before merging this.

arendjantetteroo replied:
The symfony  loaders are also seperate right now because you need the ODM/ORM compiler pass interface, so if you would want to combine the 2 we need to either depend on the fixtureBundle and extend their loader or abstract the loading from the type of database adatper you need and use injection for the ObjectManager in the load() function.

If the fixtures have proper dependency injection, injecting the objectManager in the constructor avoids needing to have the ODM/ORM distinction between the compiler pass marker and you can use just one loader with one interface. 

Abstracting this into a separate bundle makes more sense than including it in the mongodb bundle i guess. 

 I'm not sure how many people use both ORM and OMD anyway? In our project we only use the ODM, so we don't need the FixturesBundle and the assorted doctrine and doctrine cache bundles it takes with it. 

Maybe for 4.0 we could put the fixture stuff in a seperate bundle that depends or extends on the FixturesBundle and override the symfony loader from the fixturesbundle to use the same Fixture interfaces and use dependency injection for the ObjectManager so it can work with both the ORM and ODM stuff. In a newer FixtureBundle they could also move to the injected ObjectManager so you don't need it anymore in the load(). 
It would mean that people that migrate to 3.6 need to change things again for 4.0, but only if they want to combine ORM/ODM. 
It's all a bit messy...

alcaeus commented on Resources/config/mongodb.xml:
```diff
@@ -210,5 +211,9 @@
         <!-- listeners -->
         <service id="doctrine_mongodb.odm.listeners.resolve_target_document" class="%doctrine_mongodb.odm.listeners.resolve_target_document.class%" public="false" />
+
+        <service id="doctrine.mongodb.odm.symfony.fixtures.loader" class="Doctrine\Bundle\MongoDBBundle\Loader\SymfonyFixturesLoader" public="false">
```
Other services are named `doctrine_mongodb.odm...`. Please follow that convention.

alcaeus commented on Tests/Command/SymfonyFixturesLoaderMock.php:
```diff
@@ -0,0 +1,32 @@
+    {
+        return $this->fixtures;
+    }
+}
```
Missing newline here.

alcaeus commented on Tests/Command/SymfonyFixturesLoaderMock.php:
```diff
@@ -0,0 +1,32 @@
+<?php
+/**
```
Please drop this comment, it serves no purpose.

arendjantetteroo commented:
@alcaeus thanks for the review, updated and rebased. 

Thinking about the ODM/ORM stuff, the only clean way would be extracting the loading of fixtures into a seperate symfony bundle that can work with the same Fixture/FixtureGroup interface and then both the FixtureBundle and this bundle/a new ODMFixture bundle depending on this FixtureLoaderBundle. 

I just don't know if that is worth all the hassle though...

arendjantetteroo commented:
@alcaeus what's the plan, anything i can do to help?

alcaeus commented:
> @alcaeus what's the plan, anything i can do to help?

No, I just need to finish looking through this and merge it. I'll try to get that done soon, even if it means copy/pasting large parts of the ORM fixtures bundle. We should clean that up afterwards.

alcaeus commented:
@arendjantetteroo I've fixed a couple of minor review issues. From a code perspective this looks good, but I think we need to add some tests for this. In DoctrineFixturesBundle, there's an [IntegrationTest](https://github.com/doctrine/DoctrineFixturesBundle/blob/master/Tests/IntegrationTest.php) that we could base our tests on. Do you want to give this a shot or would you rather have me do it? I won't have time for it before the weekend though...

arendjantetteroo commented:
I'll see what i can do, not sure i get to it before the weekend either :)

arendjantetteroo commented:
@alcaeus I copy/pasted and then changed the IntegrationTest, works for me locally, waiting for the travis checks now. 
I added a seperate commit for the type removal so it supports 5.6 in 3.6 version of the bundle. 
This commit can then be reverted for 4 to have types if wanted. If we squash it, this info is lost.

arendjantetteroo commented:
Fails on pre 7.1 with : 
`1) Doctrine\Bundle\MongoDBBundle\Tests\IntegrationTest::testFixturesLoader
Symfony\Component\DependencyInjection\Exception\InvalidArgumentException: Unable to replace alias "doctrine.odm.mongodb.document_manager" with actual definition "doctrine_mongodb.odm.document_manager".`

Any ideas?

arendjantetteroo commented:
@alcaeus From what i understand i need to load a config into the Extension to get this document manager definition in the service container. However i don't understand why this works fine in 7.1 and 7.2?

alcaeus commented:
I'm taking a look what's going on - no idea why it isn't working. We'll see 👍

alcaeus commented:
@arendjantetteroo I inadvertently force-pushed to your remote, overwriting your commits which I wanted to keep. Can you force-push your previous branch over my changes? Thanks!

arendjantetteroo commented:
@alcaeus I got a message to force push but i think you already corrected it right?

alcaeus commented:
> @alcaeus I got a message to force push but i think you already corrected it right?

Yeah, I fixed it. Currently trying to fix the alias issue you mentioned above.
